Lib/test/test_posix.py

```diff
@@ -2194,6 +2194,59 @@ def test_utime(self):
                 os.utime("path", dir_fd=0)
 
 
+class NamespacesTests(unittest.TestCase):
+    """Tests for os.unshare() and os.setns()."""
+
+    @support.requires_subprocess()
+    def subprocess(self, code):
+        import subprocess
+        with subprocess.Popen((sys.executable, '-c', code),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            encoding="utf-8"
+        ) as p:
+            p.wait()
+            return (
+                p.returncode,
+                tuple(p.stdout),
+                tuple(p.stderr),
+            )
+
+    @unittest.skipUnless(hasattr(os, 'unshare'), 'needs os.unshare()')
+    @unittest.skipUnless(hasattr(os, 'setns'), 'needs os.setns()')
+    @unittest.skipUnless(os.path.exists('/proc/self/ns/uts'), 'need /proc/self/ns/uts')
+    @support.requires_linux_version(3, 0, 0)
+    def test_unshare_setns(self):
+        rc, _, err = self.subprocess("""if 1:
+            import os
+            import sys
+            fd = os.open('/proc/self/ns/uts', os.O_RDONLY)
+            try:
+                original = os.readlink('/proc/self/ns/uts')
+                os.unshare(os.CLONE_NEWUTS)
+                new = os.readlink('/proc/self/ns/uts')
+                if original == new:
+                    raise Exception('os.unshare failed')
+                os.setns(fd, os.CLONE_NEWUTS)
+                restored = os.readlink('/proc/self/ns/uts')
+                if original != restored:
+                    raise Exception('os.setns failed')
+            except OSError as e:
+                sys.stderr.write(str(e.errno))
+                sys.exit(2)
+            finally:
+                os.close(fd)
+            """)
+
+        if rc == 2:
+            e = int(err[0])
+            self.assertIn(e, (errno.EPERM, errno.EINVAL, errno.ENOSPC, errno.ENOSYS))
```

Comment:
I would prefer to move these checks inside the child process code. I don't think that skipping the test if unshare() or setns() fails matters. Just exit the child process early with a success (exit code 0, well, just exit) if you get one of these errors, no? If the child process must always succeed, you can use test.suppot.script_helper.assert_python_ok() which captures stdout and stderr and raises an exception with all data if the process fails.

Maybe it would make sense to add a ``except PermissionError:`` instead. Would you mind to add a comment explaining why/how PermissionError can happen?

ENOSPC is expected on unshare() if we exceed some limits on namespaces? If it can only happen on unshare(), maybe add a try/except only on this function call?

Why is EINVAL expected? Can it be raised if there is a bug in the test? Or can it be raised depending on the kernel configuration?

Reply:
Cool! did not know about `assert_python_ok`

Reply:
`EINVAL` could be raised the kernel was not configured with the `CONFIG_UTS_NS` option.